src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php

```diff
@@ -305,4 +306,193 @@ private function getRedisEvictionPolicy(): string
+        foreach ($tagKeys as $tagKey) {
+            $tagKey = substr($tagKey, strlen($prefix));
+            $cursor = null;
+            $hasExistingKeys = false;
+            do {
+                // Fetch all referenced cache keys from the tag entry.
+                $results = $this->pipeline(function () use ($getSetContentLua, $tagKey, $cursor) {
+                    yield 'eval' => [$getSetContentLua, [$tagKey, $cursor], 1];
+                });
+                [$cursor, $referencedCacheKeys] = $results->valid() ? $results->current() : [null, null];
+
+                if (!empty($referencedCacheKeys)) {
+                    // Counts how many of the referenced cache items exist.
+                    $existingCacheKeysResult = $this->pipeline(function () use ($referencedCacheKeys) {
+                        yield 'exists' => $referencedCacheKeys;
+                    });
+                    $existingCacheKeysCount = $existingCacheKeysResult->valid() ? $existingCacheKeysResult->current() : 0;
+                    $hasExistingKeys = $hasExistingKeys || ($existingCacheKeysCount > 0 ?? false);
+
+                    // If compression mode is enabled and the count between
+                    // referenced and existing cache keys differs collect the
+                    // missing references.
+                    if ($compressMode && count($referencedCacheKeys) > $existingCacheKeysCount) {
+                        // In order to create the delta each single reference
+                        // has to be checked.
+                        foreach ($referencedCacheKeys as $cacheKey) {
+                            $existingCacheKeyResult = $this->pipeline(function () use ($cacheKey) {
+                                yield 'exists' => [$cacheKey];
+                            });
+                            if ($existingCacheKeyResult->valid() && !$existingCacheKeyResult->current()) {
+                                $orphanedTagReferenceKeys[$tagKey][] = $cacheKey;
+                            }
+                        }
+                    }
+                    // Stop processing cursors in case compression mode is
+                    // disabled and the tag references existing keys.
+                    if (!$compressMode && $hasExistingKeys) {
+                         break;
+                    }
+                }
+            } while ($cursor = (int) $cursor);
+            if (!$hasExistingKeys) {
+                $orphanedTagKeys[] = $tagKey;
+            }
+        }
+
+        $stats = ['orphanedTagKeys' => $orphanedTagKeys, 'tagKeys' => $tagKeys];
+        if ($compressMode) {
+            $stats['orphanedTagReferenceKeys'] = $orphanedTagReferenceKeys;
+        }
+        return $stats;
+    }
+
+    /**
+     *
+     * @TODO Verify the LUA scripts are redis-cluster safe.
+     *
+     * @param bool $compressMode
+     * @return bool
+     */
+    protected function pruneOrphanedTags(bool $compressMode = false): bool
```

Comment:
why is this a protected method ? I don't see why we would introduced an inheritance-based extension point there, that we would have to maintain

Reply:
agreed, changed visiblity